### PR TITLE
Replace policy.submitsTo with PolicyUtils.getSubmitToAccountID

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

We want to create a new util function `PolicyUtils.getSubmitToAccountID` and replace `policy.submitsTo` usage with the new function.
### Fixed Issues

$ https://github.com/Expensify/App/issues/40356
PROPOSAL: https://github.com/Expensify/App/issues/40356#issuecomment-2062986918

### Tests

Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps
### QA Steps

A. Prevent self-approval
1. Go to collect workspace setting in OldDot
2. Enable prevent self-approval settings
3. In Workflow tab, either disable Add approval or enable it and set it to yourself
4. Go to collect workspace chat in NewDot
5. Request money
6. Verify the Submit button is disabled

Note: Remember to disable prevent self-approval back

B. Approval optional
1. Go to collect workspace setting
2. Open the Workflow tab
3. Disable the Add approvals
4. Go to collect workspace chat
5. Request money
7. Submit the request
8. Verify the submit request is successful and the pay button is shown

C. Self-approval
1. Go to collect workspace setting
2. Open the Workflow tab
3. Enable Add approval and set the approver as yourself
4. Go to collect workspace chat
5. Request money
6. Open the expense report
7. Submit the request
9. Verify the next step below the header is "Waiting for you to review these expenses"
10. Verify there is an Approve button

D. Other approver
1. Go to collect workspace setting
2. Open the Workflow tab
3. Enable Add approval and set the approver to another user
4. Go to collect workspace chat
5. Request money
6. Open the expense report
7. Submit the request
8. Verify the next step below the header is "Waiting for {another user name} to approve these expenses"
9. Verify the approve button doesn't shows
11. As the approver, go to the workspace chat and approve the request
12. Verify the approve successful
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

A. Prevent self-approval
Screenshot 2024-04-20 at 13 00 34

B. Approval optional

https://github.com/Expensify/App/assets/50919443/6e19b3a0-7fc5-4059-a76c-c4d92f68e835

C. Self-approval

https://github.com/Expensify/App/assets/50919443/ccf07fdb-cc53-4aa1-88da-63bdc9830029

D. Other approver

https://github.com/Expensify/App/assets/50919443/8429263f-62cf-4ecf-b8bf-c9a5b37c2c46

</details>

<details>
<summary>Android: mWeb Chrome</summary>

A. Prevent self-approval
Screenshot 2024-04-20 at 12 55 00

B. Approval optional

https://github.com/Expensify/App/assets/50919443/58518c39-19ee-4cbe-89f7-124910769284

C. Self-approval

https://github.com/Expensify/App/assets/50919443/53583ed0-d740-4705-8214-13d426940fc7

D. Other approver

https://github.com/Expensify/App/assets/50919443/5597c42a-1841-4a44-a38a-ef723cb53125

</details>

<details>
<summary>iOS: Native</summary>

A. Prevent self-approval
Screenshot 2024-04-20 at 12 55 54

B. Approval optional

https://github.com/Expensify/App/assets/50919443/95de2c7c-500f-44cf-a657-082f264df8d3

C. Self-approval

https://github.com/Expensify/App/assets/50919443/09a3b231-a678-47cb-bded-eb4826dd530d

D. Other approver

https://github.com/Expensify/App/assets/50919443/b4a6ad5f-a424-4a3a-93bf-36054a43e21b

</details>

<details>
<summary>iOS: mWeb Safari</summary>

A. Prevent self-approval
Screenshot 2024-04-20 at 12 54 47

B. Approval optional

https://github.com/Expensify/App/assets/50919443/2635e3d2-4cbd-430b-94e4-2e8b72cf3539

C. Self-approval

https://github.com/Expensify/App/assets/50919443/e9713eb5-e301-40e6-b078-a896ca63b764

D. Other approver

https://github.com/Expensify/App/assets/50919443/6a2d4462-3cd4-44c2-93b4-d54020f923d0

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

A. Prevent self-approval
Screenshot 2024-04-19 at 14 58 21

B. Approval optional

https://github.com/Expensify/App/assets/50919443/b49ebb6f-b22d-4ff4-bfce-c7a955a9aac2

C. Self-approval

https://github.com/Expensify/App/assets/50919443/a8b7df0b-9469-4ac5-bfed-342471db39de

D. Other approver

https://github.com/Expensify/App/assets/50919443/84b7b02c-b574-44a1-8f4f-41cf8769b71f

https://github.com/Expensify/App/assets/50919443/b883f5dd-1ce8-4b97-b37e-ab26e7f5a0b4

</details>

<details>
<summary>MacOS: Desktop</summary>

A. Prevent self-approval
Screenshot 2024-04-20 at 12 54 13

B. Approval optional

https://github.com/Expensify/App/assets/50919443/b2878733-e330-439a-8476-b791d0efd5b4

C. Self-approval

https://github.com/Expensify/App/assets/50919443/0a6dcdff-23eb-4bd8-b26c-0bf19d072b56

D. Other approver

https://github.com/Expensify/App/assets/50919443/439aa54b-3ded-4732-af04-69fb65f40044

</details>

### CODE REVIEW

**Comment by bernhardoj** [comment]
@flodnv 
1. I decided to accept accountID instead of email in `PolicyUtils.getSubmitToAccountID` param since accountID is easier to get. Inside the function, `getLoginsByAccountIDs` to get the employee email.
2. So far I tested 4 cases, prevent self-approval, approval optional, self-approval, and other approver.
3. I pass `ownerAccountID` instead of the `currentUserAccountID` for all usages, please check whether it's correct or not.
4. I only record for the web since the code doesn't depend on the platform

cc: @getusha

File: src/libs/PolicyUtils.ts
Comment (flodnv) [inline src/libs/PolicyUtils.ts:335]:
```suggestion
```

File: src/libs/PolicyUtils.ts
Comment (flodnv) [inline src/libs/PolicyUtils.ts:329]:
```suggestion

    // For policy using the optional or basic workflow, the manager is the policy default approver.
```

File: src/libs/ReportUtils.ts
Comment (flodnv) [inline src/libs/ReportUtils.ts:3434]: Please move this comment to line 3432 and rephrase it to:

```
 // Get the approver/manager for this report to properly display the optimistic data
```

**Comment by flodnv** [comment]
@bernhardoj looking good, thanks! 👍

**Comment by flodnv** [comment]
>I only record for the web since the code doesn't depend on the platform

Please always test on all platforms 🙇

**Comment by bernhardoj** [comment]
Updated the code, will complete the rest of the recording tomorrow.

**Comment by bernhardoj** [comment]
Recording/screenshot complete.

**Comment by trjExpensify** [comment]
👋 @getusha this is a deliverable for our May release, can you review the PR today please? Thanks!

File: src/libs/PolicyUtils.ts
Comment (flodnv) [inline src/libs/PolicyUtils.ts:335]: Let's add a function doc here like:

> Returns the accountID to whom the given `employeeAccountID` submits reports to in the given `Policy`.

**Comment by ikevin127** [comment]
I'm taking over the review here based on https://github.com/Expensify/App/issues/40356#issuecomment-2070414715.
♻️ Review in progress, ETA ~1h, max 1h30m.

**Comment by trjExpensify** [comment]
Assigned you, @ikevin127!

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

| A | B | C | D |
| ------------- | ------------- | ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/b0a1ee09-1074-4fb1-8bb2-98099db402b4"> | <video src="https://github.com/Expensify/App/assets/56457735/1fde1e0f-96a3-46e6-9c14-c1bf4acad891"> | <video src="https://github.com/Expensify/App/assets/56457735/45867326-71b1-45af-a69b-7689a2c417a6"> | <video src="https://github.com/Expensify/App/assets/56457735/4c790d52-89e4-4bd8-af5f-4c8d07faa8a3"> |

</details>

<details>
<summary>Android: mWeb Chrome</summary>

| A | B | C | D |
| ------------- | ------------- | ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/ae1aeccc-15d6-4f55-afd9-7d6793cbb213"> | <video src="https://github.com/Expensify/App/assets/56457735/80061cfe-1239-423b-8aab-d89458330754"> | <video src="https://github.com/Expensify/App/assets/56457735/bf4706e5-fd25-4c46-8a12-92eb12023a1d"> | <video src="https://github.com/Expensify/App/assets/56457735/140ba4fc-b9a2-42bc-9628-090585f23040"> |

</details>

<details>
<summary>iOS: Native</summary>

| A | B | C | D |
| ------------- | ------------- | ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/e8e67fa1-f284-4962-bf74-a05efd1a17a3"> | <video src="https://github.com/Expensify/App/assets/56457735/7d349203-12ce-4e3d-8aed-ad7f6a7fd114"> | <video src="https://github.com/Expensify/App/assets/56457735/1fd0d020-7c83-475f-b54d-44eb97b32776"> | <video src="https://github.com/Expensify/App/assets/56457735/fc12190b-92ed-403f-ad57-8bfacdb8edb4"> |

</details>

<details>
<summary>iOS: mWeb Safari</summary>

| A | B | C | D |
| ------------- | ------------- | ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/c5c5bffd-f6c8-4d92-8e34-6fb0f2bcbab4"> | <video src="https://github.com/Expensify/App/assets/56457735/b806de38-5bcc-48ac-bf9b-9860f11df3e1"> | <video src="https://github.com/Expensify/App/assets/56457735/c8277473-a327-4f7f-8b2c-e4e27198d7cc"> | <video src="https://github.com/Expensify/App/assets/56457735/3966e50f-f8ca-4e76-8d87-5c550587dbbf"> |

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

| A | B | C | D |
| ------------- | ------------- | ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/7e4a7ab1-f967-4ce1-9830-34f547565518"> | <video src="https://github.com/Expensify/App/assets/56457735/916ab020-0ef7-4198-b682-d116249c687c"> | <video src="https://github.com/Expensify/App/assets/56457735/4b6368bd-b56a-4ed6-913c-c10a30d91c7b"> | <video src="https://github.com/Expensify/App/assets/56457735/70a21f35-5c52-4dab-935e-ebd7c483b970"> |

</details>

<details>
<summary>MacOS: Desktop</summary>

| A | B | C | D |
| ------------- | ------------- | ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/1c779fc1-e9f0-4f58-9ade-9467cf545b17"> | <video src="https://github.com/Expensify/App/assets/56457735/48f74fff-64d2-40b0-aa16-34331c1a092c"> | <video src="https://github.com/Expensify/App/assets/56457735/42ece68f-8c87-4beb-8c4f-861f416fddb1"> | <video src="https://github.com/Expensify/App/assets/56457735/1ed9a6d2-0aef-4575-936c-4702987d4a82"> |

</details>

**Comment by ikevin127** [comment]
**Test check:**
✅ `A. Prevent self-approval`
✅ `B. Approval optional`
⚠️ `C. Self-approval`
⚠️ `D. Other approver`

> [!warning]
> @bernhardoj I found issues with test `C. Self-approval` and test `D. Other approver` , hence why I left the following checkbox unchecked from the reviewer checklist:
>> [ ] I verified tests pass on **all platforms** & I tested again on:

The issues are the following for Test C:
1. After step `7. Submit the request` -> the next step below the header is "Waiting for **you** to **review** these expenses." but if we wait a bit it changes to "These expenses are scheduled to **automatically submit on Sunday**! No further action required!" which is what we usually see when we still have the `Submit` button showing, but it's showing that header while we have the `Approve` button displayed.

2. If we press the `Approve` button ->  the next step below the header is "Waiting for **you** to **pay** these expenses." but if we wait a bit it changes to "No further action required!".

To see the behaviour more clearly here's the video from web:

https://github.com/Expensify/App/assets/56457735/fe01bec6-9c48-4f16-970f-4babcd4cb8c7

> [!important]
> Same as for test C, test D has similar behaviour in the fact that if we wait a bit after we press `Submit` and `Approve` buttons, the header changes instead of displaying the correct message consistently.
>
> **Note:** for test D the message also changes after the approver opens the report if we wait a bit from "These expenses are scheduled to **automatically submit on Sunday**! No further action required!" to "**kevin.bader96+3@gmail.com** is waiting for **you** to **review** these expenses." -> then once `Approve` is pressed, the same thing as for test C happens if we wait a bit.

If this is expected behaviour let me know and I'll check the last checkbox and approve!

cc @flodnv @trjExpensify

**Comment by trjExpensify** [comment]
>After step 7. Submit the request -> the next step below the header is "Waiting for you to review these expenses." but if we wait a bit it changes to "These expenses are scheduled to automatically submit on Sunday! No further action required!" which is what we usually see when we still have the Submit button showing, but it's showing that header while we have the Approve button displayed.

>Same as for test C, test D has similar behaviour in the fact that if we wait a bit after we press Submit and Approve buttons, the header changes instead of displaying the correct message consistently.

That doesn't sound right. It shouldn't be reverting back to the previous next step. Is there a mismatch of what the OldDot next steps banner shows when this happens if you look at the corresponding expense report? 

CC: @mountiny as perhaps being related to "optimistic next steps"

**Comment by ikevin127** [comment]
> Is there a mismatch of what the OldDot next steps banner shows when this happens if you look at the corresponding expense report?

Just tested and there's no mismatch on OD side, only thing is that on OD we have to refresh the page to see the changes but the next step status is correct on OD compared to issues reported from testing on this branch.

I performed a `npm ci` again to make sure there's no cached code or anything like that.
I did a re-test and now test `C. Self-approval` seems to pass, meaning it matches OD behaviour:

https://github.com/Expensify/App/assets/56457735/9ecc34f0-40e5-4327-b4ce-26a793346e57

Now only test `D. Other approver` seems to behave weirdly, after the other approver navigates to the report:

https://github.com/Expensify/App/assets/56457735/e765f7df-2c61-4f04-974e-0bcf7c642604

> [!note]
> If this issues are not reproducible on anybody else's side, neither author or CME, then I can approve as there might be something going on specifically on my side which causes the weird behaviour. Let me know!

**Comment by bernhardoj** [comment]
So, for test D, if I understand correctly, there are 2 issues:
1. When the other approver opens the submitted report, the next step still shows "These expenses are scheduled to automatically submit on Sunday! No further action required!" for a while. This is a BE problem where the pusher doesn't send the updated `reportNextStep` data.

The outdated next step in storage
Screenshot 2024-04-23 at 10 43 28

The updated next step received from OpenReport
Screenshot 2024-04-23 at 10 45 16

2. When the other approver approves the request, the next step shows "Waiting for you to pay these expenses" even though they are not the ones who pay the expense, and then the message changes to "No further action required!" which is the correct one.

The optimistic next step for approved action doesn't have a case when to show the "No further action required" case
https://github.com/Expensify/App/blob/64695c850dfd7dc054823b69f35c9511cd934cdd/src/libs/NextStepUtils.ts#L273-L322

This issue happens on the staging too.

https://github.com/Expensify/App/assets/50919443/6588563b-385e-449d-84e8-1107bb98189c

**Comment by bernhardoj** [comment]
For the test C, I can reproduce what you reproduced earlier @ikevin127. Here is the root cause of each issue:
1. After submitting the request, we optimistically set the next step to "Waiting for you to review these expenses.", but the OpenReport request will return the next step as "These expenses are scheduled to automatically submit on Sunday! No further action required!". 

How can this happen? If you open the report and quickly submit it before the OpenReport request completes, the outdated response from OpenReport will overwrite our optimistic next step.

2. The ApproveMoneyRequests response sets the next step to "No further action required!". I think this is a BE issue because we still need to pay the expense.

Both happens on main too.

**Comment by ikevin127** [comment]
@bernhardoj Thanks for checking!

Indeed for Test C it's on and off for me regarding the issue, sometimes it works as expected, sometimes next steps text jumps back.

As for Test D, if that happens on staging too and CME is fine with moving on in this state then that's fine with me.

@flodnv @trjExpensify Let me know and I'll complete the checklist and approve!

**Comment by mountiny** [comment]
> The optimistic next step for approved action doesn't have a case when to show the "No further action required" case

This is probably expected, I believe that we did not have the dedicated reimburser logic yet when the next steps were built optimistically

Eitherway, what is coming from the BE should be correct after taking some action so we should make sure the optimistic steps match it. They did match it when it was build first time but we are adding more policy data/ conditions to the app throughout simplified collect so it could be that some conditions should have been updated along the way too

Dont forget there are unit tests for the next steps so please cover everything that is changing there

**Comment by trjExpensify** [comment]
What are the next steps here?

**Comment by bernhardoj** [comment]
We are not changing the next step in this PR, but just to make sure the next step still works as expected, but @ikevin127 found some issues with it that also exist on `main`. I think we need @flodnv confirmation on what to do with all of those issues. The root cause is explained [here](https://github.com/Expensify/App/pull/40532#issuecomment-2071314785) and [here](https://github.com/Expensify/App/pull/40532#issuecomment-2071324283)

**Comment by bernhardoj** [comment]
I agree with that

cc: @ikevin127

**Review by ikevin127 — APPROVED** [review]
Thanks everybody for chiming-in on this!

Based on the latest discussions it looks like we're fine moving on with this PR even though there were some issues with two of the test scenarios during testing, issues related to the next steps logic / incoming BE data.

These will be handled as separete issues and therefore won't be reported as regressions of this PR post-merge, that's all I wanted to confirm!

:shipit:

**Comment by flodnv** [comment]
Thanks @ikevin127 -- can you please report these in #expensify-bugs ?

**Review by flodnv — APPROVED** [review]

**Comment by ikevin127** [comment]
> Thanks @ikevin127 -- can you please report these in #expensify-bugs ?

@flodnv Reported in [Slack 🧵](https://expensify.slack.com/archives/C049HHMV9SM/p1714080458273049), feel free to add more info in 🧵 if you think I missed anything.

**Comment by eh2077** [comment]
@bernhardoj @ikevin127  Seems like this PR caused https://github.com/Expensify/App/issues/41081, is it a regression of this PR or we should handle it separately?

Cc @flodnv

**Comment by bernhardoj** [comment]
It's an [existing issue](https://github.com/Expensify/App/pull/40532#issuecomment-2071314785), so should be handled separately

**Comment by flodnv** [comment]
@bernhardoj in [this PR](https://github.com/Expensify/App/pull/39321#pullrequestreview-2014687591) we added another usage of `policy.submitsTo` here: https://github.com/Expensify/App/pull/39321/files#diff-577fcc5f3a5e4930916ff310e61c38edfa093792fe3ab55b9a30d2aecd8811dbR3487, 2 days before we merged this PR 😞 Can you please update that code to the newer version?

**Comment by bernhardoj** [comment]
@flodnv Sure, here is the [PR](https://github.com/Expensify/App/pull/43019)

**Pull request merged**
